src/realm_dart.cpp

```diff
@@ -87,5 +87,11 @@ RLM_API void realm_dart_userdata_async_free(void* userdata)
     });
 }
 
+RLM_API void realm_dart_invoke_unlock_callback(bool success, void* unlockFunc)
+{
+    auto castFunc = (reinterpret_cast<realm::util::UniqueFunction<void(bool)>*>(unlockFunc));
+    (*castFunc)(success);
+}
```

Comment:
I dislike that this is now so specific. There's nothing in the function implementation that indicates that this is somehow related to locking. It's just invoking a function that takes a boolean. The fact that in the client reset flow it'll unlock a lock is an implementation detail.

Reply:
This exactly a specific function to unlock and the argument is called unlockFunc, hence the otherway around was weird. This makes it more clear what it does and when it should be used.

Reply:
I think we can renamed only the argument name. I has left unchanged after I refactored the function to be general.

Reply:
I also think we should rename the argument - it's a void* which is then cast to `function(bool)`. Nothing about it deals with unlocking anything.

Reply:
It just call a function. You specify the pointer of that function as an argument. The method doesn't care what is the function just calls it. The specific this here is only that the function expects a bool argument. It could be renamed to `realm_dart_invoke_function_with_confirmation`

Reply:
the argument just names what the void* pointer points to. The name of the function names what it does.

Reply:
The `void*` pointer points to a `Function(bool)`. How is that related to locks?

Reply:
it specifies that this function `realm_dart_invoke_unlock_callback` works with specific native callbacks expected to be passed as an argument.  You can pass any other function but that will be wrong usage and we can not express that other than the parameter name

Reply:
I disagree with this - it's a function that executes a native function. What the native function does is irrelevant as long as it takes a single boolean argument.

Reply:
We are not trying to implement a generic function that can call native functions with a bool argument. We are writing a very specific function that should do only one thing and is clear what it does and when it should be used. We don't have a case for anything else. 
If we are trying to make a generic function then the questions will be why this signature and not a different one. Why support bools and not even more. Then why don't we stop right now and make 100 functions with different generic meaning.
This is not useful since we don't have such case. Hence we should be explicit here when this function should be used.

Reply:
I'm not sure I subscribe to this ideology. While I agree we're not trying to implement a general-purpose function calling mechanism, I firmly believe we should name functions for what they do and not for how they're being used. The name of the function as is implies that it somehow manages locking/unlocking of resources, but it doesn't do it - it just invokes another function with a boolean argument.

Reply:
If it sounds to you better to be specific, why did you move it realm_dart.cpp? Let's specify it for reset callbacks and to leave it next to the locking functions in realm_dart_sync.cpp

Reply:
because it can be used for other invocations that expect a result synchronously as well. it is not tied to sync specifically.

Reply:
@nikola This function is named after what it does and not how it is used. It invokes a unlock function callback which was passed to a Dart function which accepts a unlock callback argument. This ties the whole concept together. So later when someone needs to do the same it is obvious what needs to be done. create a static dart function which should return a result using that function and continue the operation

Reply:
Again, the function being passed as an argument is not an unlock function. It's a function(bool).

Reply:
Consider this extreme example:

```dart
void bark(Object dog) {
  final animal = dog as Animal;
  animal.produceSound();
}

final dog = Dog();
dog.bark();
```

Do you think the bark function is named correctly here? Because this is what is essentially happening with this C function. We accept a generic void* pointer, we cast it to a function(bool), then invoke that function. And we call the argument `unlockFunc` because we expect that consumers are going to invoke us with a function that unlocks something.

Reply:
I don't think we are getting anywhere here. I would like to move away from bike shading the function name. The function is called `invoke_unlock_callback` and it does what it says and expect a specific argument to be passed in and be used in a specific case. Lets move the discussion to another place where we can gain more value.
The other feedback for deadlocks is more important in this PR. I am trying to handle it.

Reply:
I'm fine with moving the discussion elsewhere, but I disagree with changing the function name here. Since both Desi and I believe the old name is better and the name change is not fixing an actual bug, let's keep the original name and we can rename it once we've completed bikeshedding.